### ISSUE TITLE
CoinJoin Integration

### ISSUE THREAD

**Issue by nixbitcoin** [fort-nix/nix-bitcoin#32]
**Time Horizon:** Medium Term

**Description:** Integrate a CoinJoin service into nix-bitcoin.

**Motivation:** Running a CoinJoin service 24/7 is desirable to maximize probability of success. Users may not wish to keep their laptops on 24/7, but nix-bitcoin is designed to be on 24/7. So it makes sense to run a CoinJoin service on a nix-bitcoin node.

**Possible Implementation:**

Option 1: Integrate [JoinMarket](https://github.com/JoinMarket-Org/joinmarket-clientserver) maker as well as taker, using its command-line UI over ssh.

Option 2: Integrate [Wasabi](https://github.com/zkSNACKs/WalletWasabi) Daemon as-is by pushing users walletfile to nix-bitcoin node as a temporary file with `nixops deploy` and starting a mix on the nix-bitcoin node (`wassabee --mix --wallet <WalletName>`). When the user wishes to stop the mix, nix-bitcoin stops the `wassabee --mix` and automatically deletes the walletfile. When the user starts up his Wasabi GUI, it should display the now post-mix coins.

Possible problems
* User going online while mix is ongoing on nix-bitcoin node
* User stopping `wassabee --mix` in the middle of a round, possibly causing DOS punishments
* Wasabi GUI not being able to accurately display new state after `wassabee --mix` on the nix-bitcoin node. This could possibly be fixed by copying the walletfile back to the users computer after he stops the mix.

Option 3: Wait for a fully-featured Wasabi headless daemon (receive, mix and send) and integrate that

Possible problems
* Both Option 2 and Option 3 (ab)use Wasabi as an intermediary wallet for CoinJoining

Option 4: Wait for and integrate Samourai's whirlpool daemon

**Comment by SamouraiDev** [comment]
Samourai whirlpool daemon and GUI are ready for release. Ping us.

**Comment by nopara73** [comment]
Hi there, I'll share some of my thoughts on the specific issues you brought up and I'll bring up other ones, too.

I'd say let me know if there's something we can implement, but I won't. Currently we ended up adding too many features, which makes the wallet less reliable and more cluttery. Because of this we had to commit to a "feature freeze" to fix up all the instabilities for the time being: https://github.com/zkSNACKs/WalletWasabi/issues/1412

Also we have great plans to completely refactor the mixing process based on the latest research on this topic, so adding daemon options are likely to be changed at this point. However I the current daemon functionality will never go away. I feel very strongly about not breaking the user space.

Quick note. IMO users should use the `--keepalive` option so when everything is mixed to the desired anonymity target, then Wasabi is waiting if new coins arrive: `wassabee --mix --wallet <WalletName> --keepalive`.

> User going online while mix is ongoing on nix-bitcoin node

You mean offline? Why is this an issue?

> User stopping wassabee --mix in the middle of a round, possibly causing DOS punishments

For one, the user can stop the daemon with ctrl+c, but the daemon won't stop until the round is finished.
Also note that Wasabi was designed for constantly failing rounds. Although the last release we barely have failing rounds, because of honest users, which is encouraging, anyway it's outside the scope of this.  
The bottom line is, I am the least worried about daemon users DoS attacking. GUI users are the more "interactive" ones: )  

> Wasabi GUI not being able to accurately display new state after wassabee --mix on the nix-bitcoin node. This could possibly be fixed by copying the walletfile back to the users computer after he stops the mix.

Yeah, it can't unless we build a fully featured RPC daemon, which we are not planning anytime soon.  

> Option 3: Wait for a fully-featured Wasabi headless daemon (receive, mix and send) and integrate that

I believe if the current functionality is insufficient, which probably is I'd prefer this solution, since it's the cleanest way to do this.

**Comment by nopara73** [comment]
Also note that, mixing with the same wallet simultaneously can be problematic. At least it was in the past, yet I would not like to provide guarantees for such behavior: https://medium.com/@molnardavid84/wasabireport-704616de55d1

**Comment by nixbitcoin** [comment]
> You mean offline? Why is this an issue?

I meant going online on his desktop GUI while mixes are running with the daemon on the nix-bitcoin node. Similar to what you said about "mixing with the same wallet simultaneously" being problematic.

> For one, the user can stop the daemon with ctrl+c, but the daemon won't stop until the round is finished.

Good to know we can at least always stop Wasabi daemon safely through systemd.

> Yeah, it can't unless we build a fully featured RPC daemon, which we are not planning anytime soon.

That would be a solution: building a Wasabi Daemon which always runs on the nix-bitcoin node and using the Wasabi desktop GUI to control it. I realize this is not the main focus of Wasabi, but it would be a good solution and we (@fort-nix) can wait to integrate this in the future.
